File: typescript/packages/bench/src/exporters/markdownExporter.ts
/** Pipe table with a header rule. */

import type { ReportRow } from "../contracts/reportContract";
import { AbstractExporter } from "../support/abstractExporter";

/** Pipe table with a header rule. */
export class MarkdownExporter extends AbstractExporter {
    constructor() {
        super("md", "text/markdown");
    }

    override export(rows: readonly ReportRow[]): string {
        const head = ["| label | amount |", "| --- | --- |"];
        const body = rows.map((row) => `| ${row.label} | ${this.formatCents(row.cents)} |`);

        return [...head, ...body].join("\n");
    }
}

File: rust/crates/atelier-bench/src/exporters/markdown_exporter.rs
//! Pipe table with a header rule.

use crate::contracts::exporter_contract::ExporterContract;
use crate::contracts::formatter_contract::FormatterContract;
use crate::contracts::report_contract::ReportRow;
use crate::support::abstract_exporter::ExporterBase;

/// Pipe table with a header rule.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct MarkdownExporter {
    base: ExporterBase,
}

impl MarkdownExporter {
    pub const fn new() -> Self {
        Self {
            base: ExporterBase::new("md", "text/markdown"),
        }
    }

    /// Shared extension/mime pair and the cent formatter.
    pub const fn base(&self) -> &ExporterBase {
        &self.base
    }
}

impl Default for MarkdownExporter {
    fn default() -> Self {
        Self::new()
    }
}

impl ExporterContract for MarkdownExporter {
    fn extension(&self) -> &'static str {
        self.base.extension
    }

    fn mime(&self) -> &'static str {
        self.base.mime
    }

    fn export(&self, rows: &[ReportRow]) -> String {
        let mut out = vec![
            "| label | amount |".to_string(),
            "| --- | --- |".to_string(),
        ];
        out.extend(
            rows.iter()
                .map(|row| format!("| {} | {} |", row.label, self.base.format_cents(row.cents))),
        );

        out.join("\n")
    }
}
